pkgs/build-support/trivial-builders.nix

```
@@ -1,5 +1,31 @@
-{ lib, stdenv, stdenvNoCC, lndir, runtimeShell }:
+{ lib, stdenv, stdenvNoCC, lndir, runtimeShell, zsh, dash, bash}:
 
+let
+  runtime = { 
+    preamble = "#!${runtimeShell}";
+    checkPhase = ''
+      ${stdenv.shell} -n $out
+    '';
+  };
+  bash = { 
+    preamble = "#!${lib.getBin bash}/bin/bash"; 
+    checkPhase = ''
+      ${lib.getBin bash}/bin/bash -n $out
+    '';
+  };
+  zsh = { 
+    preamble = "#!${lib.getBin zsh}/bin/zsh"; 
```

Could do

```suggestion
    preamble = "#!${lib.getBin zsh}/${zsh.shellPath}"; 
```

and same with check path below. Indeed could generalize this by creating a function that takes the shell package as an argument.

Hmm, better yet. Maybe it would be possible to put the `write*` functions directly as a passthrough in the derivation? It would be pretty cool if could do something like

```nix
pkgs.zsh.writeProgram "foo" {} ''
  echo hello world
''
```

or

```nix
pkgs.fish.writeProgram "foo" {} ''
  echo hello world
''
```

or even

```nix
pkgs.ghc.writeProgram "foo" {} ''
  main = putStrLn "hello world"
''
```

I kind of like that since it feels more modular, we wouldn't have to put so many new functions in trivial-builders. It would also be possible to more naturally use a specific package (e.g., with overrides) in the creation of the program.

---

Note, my idea with the `{}` is that it would allow whatever customization that may be necessary, e.g.,

```nix
pkgs.zsh.writeProgram "foo" { executable = false; } ''
  echo hello world
''
```

---

Created a quick and dirty proof of concept in #124080.

---

I like it! It definitely makes more sense once you start thinking of `writePythonScript`, `writeJuliaScript`, etc.

I've seen more chatter lately of cleaning up nixpkgs in general (e.g. https://github.com/NixOS/nixpkgs/issues/107539), so maybe this could part of that discussion?

I'd love if there was a standardized interface for where these kinds of functions should live. I kind of like the idea of there being a canonical `langPlatform` (e.g. `rustPlatform`) for every language where all things related to `lang` could live.